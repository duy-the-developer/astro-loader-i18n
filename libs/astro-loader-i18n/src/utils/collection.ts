export function getAllUniqueKeys(obj: Record<string, unknown>, keys = new Set<string>()): Set<string> {
  Object.entries(obj).forEach(([key, value]) => {
    keys.add(key);
    if (value && typeof value === "object" && !Array.isArray(value)) {
      getAllUniqueKeys(value as Record<string, unknown>, keys);
    }
  });
  return keys;
}

export function convertToSingleLocale(obj: Record<string, unknown>, locales: string[], locale: string): Record<string, unknown> {
  return Object.entries(obj).reduce(
    (result, [key, value]) => {
      if (key === locale) {
        Object.assign(result, value);
      } else if (!locales.includes(key)) {
        result[key] =
          typeof value === "object" && !Array.isArray(value) ? convertToSingleLocale(value as Record<string, unknown>, locales, locale) : value;
      }
      return result;
    },
    {} as Record<string, unknown>
  );
}
