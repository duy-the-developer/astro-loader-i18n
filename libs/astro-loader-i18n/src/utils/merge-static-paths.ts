import { GetStaticPathsResult } from "astro";

export function mergeStaticPaths(...staticPaths: GetStaticPathsResult[]): GetStaticPathsResult {
  if (staticPaths.some((paths) => paths.length !== staticPaths[0].length)) {
    throw new Error("All static paths must have the same length");
  }

  return staticPaths[0].map((entry, index) => {
    return staticPaths.reduce(
      (merged, current) => ({
        params: { ...merged.params, ...current[index].params },
        props: { ...merged.props, ...current[index].props },
      }),
      entry
    );
  });
}
