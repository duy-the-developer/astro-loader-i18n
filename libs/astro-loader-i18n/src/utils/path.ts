const TRIM_SLASHES_PATTERN = /^\/|\/$/g;
const DOUBLE_POINTS_PATTERN = /\.{2}/g;
const DOUBLE_SLASH_PATTERN = /\/{2}/g;
const INDEX_COMMON_TRANSLATION_ID = "index";

export function joinPath(...paths: Array<string | number | undefined>) {
  return paths.filter(Boolean).join("/");
}

export function resolvePath(...paths: Array<string | number | undefined>) {
  return `/${joinPath(...paths)}`;
}

export const trimSlashes = (path: string) => {
  return path === "/" ? path : path.replace(TRIM_SLASHES_PATTERN, "");
};

export const parseLocale = (path: string, locales: string[], defaultLocale: string) => {
  const locale = locales.find((locale) => path.indexOf(locale) !== -1);

  if (locale) {
    return locale;
  }

  return defaultLocale;
};

export const createTranslationId = (path: string, pathLocale: string) => {
  path = path.replace(pathLocale, "");
  path = path.replace(DOUBLE_POINTS_PATTERN, ".");
  path = path.replace(DOUBLE_SLASH_PATTERN, "/");
  path = trimSlashes(path);

  if (path === "" || path === "/") return INDEX_COMMON_TRANSLATION_ID;

  return path;
};
