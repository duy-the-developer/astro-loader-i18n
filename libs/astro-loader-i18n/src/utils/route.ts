import { joinPath } from "./path";

type Segments = Record<string, string | undefined>;
export type SegmentTranslations = Record<string, Segments>;
type RouteSegment = {
  spread: boolean;
  param: boolean;
  value: string;
};
type RoutePattern = RouteSegment[];

/**
 * Parses a route pattern into an array of route segments.
 *
 * @param routePattern is a string that templates the a route (e.g. /[...locale]/blog/[slug])
 */
export function parseRoutePattern(routePattern: string): RoutePattern {
  const segments = routePattern
    .split("/")
    .filter((segment) => segment !== "")
    .map((segment) => {
      if (segment.startsWith("[") && segment.endsWith("]")) {
        return {
          spread: segment.startsWith("[..."),
          param: true,
          value: segment.replace("[", "").replace("...", "").replace("]", ""),
        };
      }

      return {
        spread: false,
        param: false,
        value: segment,
      };
    });

  if (segments.length === 0) {
    return [{ spread: false, param: false, value: "/" }];
  }

  return segments;
}

/**
 *
 * @param routePattern is an array consisting of route segments
 * @param segmentValues is an object containing the values for each route segment
 */
export function buildPath(routePattern: RoutePattern, segmentValues: Segments) {
  return joinPath(
    ...routePattern.map((segment) => {
      if (segment.param) {
        if (!segmentValues[segment.value] && !segment.spread) {
          throw new Error(`No segment value found for route segment "${segment.value}". Did you forget to provide it?`);
        }

        return segmentValues[segment.value];
      }
      return `${segment.value}`;
    })
  );
}
