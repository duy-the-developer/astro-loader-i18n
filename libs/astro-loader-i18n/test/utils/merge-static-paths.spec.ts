import { GetStaticPathsResult } from "astro";
import { describe, expect, it } from "vitest";
import { mergeStaticPaths } from "../../src/utils/merge-static-paths";

const fixture: GetStaticPathsResult[] = [
  [
    { params: { locale: "en" }, props: { locale: "en" } },
    { params: { locale: "de" }, props: { locale: "de" } },
  ],
  [
    { params: {}, props: { commonTransaltionId: "some/page.mdx" } },
    { params: {}, props: { commonTransaltionId: "some/page.mdx" } },
  ],
];

describe("mergeStaticPaths", () => {
  it("should merge static paths", () => {
    const merged = mergeStaticPaths(...fixture);
    expect(merged).toMatchSnapshot();
  });
});
