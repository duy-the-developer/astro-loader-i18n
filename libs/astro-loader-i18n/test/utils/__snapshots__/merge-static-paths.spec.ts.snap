// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mergeStaticPaths > should merge static paths 1`] = `
[
  {
    "params": {
      "locale": "en",
    },
    "props": {
      "commonTransaltionId": "some/page.mdx",
      "locale": "en",
    },
  },
  {
    "params": {
      "locale": "de",
    },
    "props": {
      "commonTransaltionId": "some/page.mdx",
      "locale": "de",
    },
  },
]
`;
